### PR TITLE
[Demangler] Make Node::addChild(NULL, ...) always assert.

### PR DESCRIPTION
**Pull request by al45tair**
We need this in order to gather relevant information when this situation arises; it's clearly happening (e.g. rdar://86071019), but not often enough for it to trigger in a build with assertions enabled.

rdar://89139049

### CODE REVIEW

**Review by mikeash — APPROVED** [review]
LGTM. Would be worth considering if there's any additional info to log here that might be helpful (like the parent's kind, child count, other contents?). Since the turnaround time on this is not going to be real quick.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
> LGTM. Would be worth considering if there's any additional info to log here that might be helpful (like the parent's kind, child count, other contents?). Since the turnaround time on this is not going to be real quick.

Good plan. I've made the `failAssert()` function dump the node that failed the assertion, which should supply plenty of extra information.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

File: lib/Demangling/Demangler.cpp [hunk 1]
```diff
@@ -93,8 +93,12 @@ static bool isRequirement(Node::Kind kind) {
 
 void swift::Demangle::failAssert(const char *file, unsigned line,
                                  NodePointer node, const char *expr) {
-  fprintf(stderr, "%s:%u: assertion failed for Node %p: %s", file, line, node,
+  fprintf(stderr, "%s:%u: assertion failed for Node %p: %s\n", file, line, node,
```
Comment (mikeash) [inline lib/Demangling/Demangler.cpp:96]: We should use `swift_reportError` and `fatalError` to get these messages into the crash log too.

Comment (al45tair) [reply lib/Demangling/Demangler.cpp:96]: Hmmm. That might be slightly trickier, in that `node->dump()` isn't going to send its output to the right place. It's not a bad idea to get something in there though; I'll have a look how that all works.

Comment (al45tair) [reply lib/Demangling/Demangler.cpp:96]: It isn't quite as simple as I'd like; we can't call `swift_reportError` or `fatalError` without linking `swiftRuntime`, which not everything that includes the demangler code does.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please test

**Pull request merged**